contracts/YieldStrategyLever.sol

```diff
@@ -0,0 +1,155 @@
+            pool.base().balanceOf(address(this))
+        );
+        // Mint LP token & deposit to strategy
+        pool.mintWithBase(
+            address(strategies[ilkId]),
+            msg.sender,
+            uint256(
+                pool.buyFYTokenPreview(uint128((baseReceived + baseAmount) / 3))
+            ), // TODO: what should be the fyTokenToBuy
+            0,
+            type(uint256).max
+        );
+
+        // Mint strategy token
+        uint256 tokensMinted = strategies[ilkId].mint(
+            address(LADLE.joins(ilkId))
+        );
+
+        // Borrow fyToken to repay the flash loan
+        LADLE.pour(
+            vaultId,
+            address(this),
+            int128(uint128(tokensMinted)),
+            int128(uint128(borrowAmount))
+        );
+    }
+
+    /// @param ilkId The id of the ilk being invested.
+    /// @param seriesId The seriesId corresponding to the vault.
+    /// @param vaultId The vault to repay.
+    /// @param borrowAmountPlusFee The amount of fyDai/fyUsdc that we have borrowed,
+    ///     plus the fee. This should be our final balance.
+    /// @param ink The amount of collateral to retake.
+    /// @param art The debt to repay.
+    ///     slippage.
+    function repay(
+        bytes6 ilkId,
+        bytes12 vaultId,
+        bytes6 seriesId,
+        uint128 borrowAmountPlusFee,
+        uint128 ink,
+        uint128 art
+    ) internal override {
+        IPool pool = IPool(LADLE.pools(seriesId));
+        // Payback debt to get back the underlying
+        LADLE.pour(
+            vaultId,
+            address(strategies[ilkId]),
+            -int128(ink),
+            -int128(art)
+        );
+
+        // Burn strat token to get LP
+        strategies[ilkId].burn(address(pool));
+
+        // Burn LP to get base & fyToken
+        (, uint256 bases, uint256 fyTokens) = pool.burn(
+            address(pool),
+            address(this),
+            0,
+            type(uint256).max
+        );
+        // buyFyToken
+        pool.buyFYToken(
+            address(this),
+            borrowAmountPlusFee - uint128(fyTokens),
+            uint128(bases)
+        );
+    }
+
+    /// @notice Close a vault after maturity.
+    /// @param ilkId The id of the ilk.
+    /// @param vaultId The ID of the vault to close.
+    /// @param ink The collateral to take from the vault.
+    /// @param art The debt to repay. This is denominated in fyTokens
+    function close(
+        bytes6 ilkId,
+        bytes12 vaultId,
+        uint128 ink,
+        uint128 art
+    ) internal override {
+        IStrategy strategy = strategies[ilkId];
+        IPool pool = strategy.pool();
+
+        LADLE.close(vaultId, address(strategy), -int128(ink), -int128(art));
+        // Burn Strategy Tokens and send LP token to the pool
+        strategy.burn(address(pool));
+        // Burn LP token to obtain base to repay the flash loan
+        pool.burnForBase(address(this), 0, type(uint256).max);
```

Comment:
This looks risky, could the pool not have enough base for this? In that case the user wouldn't be able to undo his position.

Reply:
Also, slippage bounds.